File: app/src/main/java/com/taipei/yanghaobo/kunu/ui/API/DogAPIFragment.java
package com.taipei.yanghaobo.kunu.ui.API;

import android.arch.lifecycle.ViewModelProviders;
import android.content.Context;
import android.net.ConnectivityManager;
import android.net.NetworkInfo;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.design.button.MaterialButton;
import android.support.design.widget.TextInputEditText;
import android.support.design.widget.TextInputLayout;
import android.support.v4.app.Fragment;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.text.Editable;
import android.text.TextUtils;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.inputmethod.InputMethodManager;

import com.taipei.yanghaobo.kunu.InjectorUtils;
import com.taipei.yanghaobo.kunu.R;


/**
 * 介接 Google APIs 用！
 */
public class DogAPIFragment extends Fragment {

    private RecyclerView mBookQueryResultRV;
    private TextInputLayout mTextInputSearch;
    private TextInputEditText mEditInputTextSearch;
    private MaterialButton mBtSearch;

    public DogAPIFragment() {
        // Required empty public constructor
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        // Inflate the layout for this fragment
        View view = inflater.inflate(R.layout.fragment_google_api, container, false);

        mBookQueryResultRV = view.findViewById(R.id.rv_book_query_result);
        mTextInputSearch = view.findViewById(R.id.text_input_search);
        mEditInputTextSearch = view.findViewById(R.id.edit_input_text_search);
        mBtSearch = view.findViewById(R.id.bt_search);

        return view;
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        BookListedAdapter adapter = new BookListedAdapter(getContext());
        mBookQueryResultRV.setAdapter(adapter);
        mBookQueryResultRV.setLayoutManager(new LinearLayoutManager(getContext()));
        BookListedVMFactory factory = InjectorUtils.provideBookListedVMFactory(getContext());
        BookListedViewModel bookListedViewModel = ViewModelProviders.of(this, factory)
                .get(BookListedViewModel.class);
        bookListedViewModel.getBookQueryResult().observe(this, books -> adapter.swapResult(books));

        mBtSearch.setOnClickListener(bt -> {
            InputMethodManager inputMethodManager = (InputMethodManager) getActivity().getSystemService(Context.INPUT_METHOD_SERVICE);
            if (inputMethodManager != null) {
                inputMethodManager.hideSoftInputFromWindow(
                        view.getWindowToken(),
                        InputMethodManager.HIDE_NOT_ALWAYS);
            }

            // 檢察網路連線
            ConnectivityManager connMgr = (ConnectivityManager) getActivity().getSystemService(Context.CONNECTIVITY_SERVICE);
            NetworkInfo networkInfo = null;
            if (connMgr != null) {
                networkInfo = connMgr.getActiveNetworkInfo();
            }
            if (networkInfo != null && networkInfo.isConnected()) {
                if (isQueryValid(mEditInputTextSearch.getText())){
                    mTextInputSearch.setError(null);
                    mTextInputSearch.setHint(getString(R.string.query_google_book));
                    // Network query
                    bookListedViewModel.queryBookListed(mEditInputTextSearch.getText().toString());
                }
                else {
                    mTextInputSearch.setError(getString(R.string.query_error));
                    mTextInputSearch.setHint(getString(R.string.input_something));
                }
            }
            else {
                mTextInputSearch.setError(getString(R.string.error_no_internet));
                mTextInputSearch.setHint(getString(R.string.error_hint_no_internet));
            }

        });
    }

    private boolean isQueryValid(Editable editable){
        if (TextUtils.isEmpty(editable)){
            return false;
        }
        return true;
    }
}

File: app/src/main/java/com/taipei/yanghaobo/kunu/ui/api/DogAPIFragment.kt
package com.taipei.yanghaobo.kunu.ui.api

import android.content.Context
import android.net.ConnectivityManager
import android.net.NetworkInfo
import android.os.Bundle
import android.text.Editable
import android.text.TextUtils
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.view.inputmethod.InputMethodManager
import androidx.fragment.app.Fragment
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProviders
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView

import com.google.android.material.button.MaterialButton
import com.google.android.material.textfield.TextInputEditText
import com.google.android.material.textfield.TextInputLayout
import com.taipei.yanghaobo.kunu.InjectorUtils
import com.taipei.yanghaobo.kunu.R
import com.taipei.yanghaobo.kunu.db.Book

/**
 * 介接 Google APIs
 */
class DogAPIFragment : Fragment() {

  private var mBookQueryResultRV: RecyclerView? = null
  private var mTextInputSearch: TextInputLayout? = null
  private var mEditInputTextSearch: TextInputEditText? = null
  private var mBtSearch: MaterialButton? = null

  override fun onCreateView(
    inflater: LayoutInflater, container: ViewGroup?,
    savedInstanceState: Bundle?
  ): View? {
    // Inflate the layout for this fragment
    val view = inflater.inflate(R.layout.fragment_google_api, container, false)

    mBookQueryResultRV = view.findViewById(R.id.rv_book_query_result)
    mTextInputSearch = view.findViewById(R.id.text_input_search)
    mEditInputTextSearch = view.findViewById(R.id.edit_input_text_search)
    mBtSearch = view.findViewById(R.id.bt_search)

    return view
  }

  override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
    super.onViewCreated(view, savedInstanceState)

    val adapter = BookListedAdapter(context)
    mBookQueryResultRV!!.adapter = adapter
    mBookQueryResultRV!!.layoutManager = LinearLayoutManager(context)
    val factory = InjectorUtils.provideBookListedVMFactory(context!!)
    val bookListedViewModel = ViewModelProviders.of(this, factory)
      .get(BookListedViewModel::class.java)
    bookListedViewModel.bookQueryResult.observe(
      this,
      Observer<List<Book>> { adapter.swapResult(it) })

    mBtSearch!!.setOnClickListener { bt ->
      val inputMethodManager = activity!!
        .getSystemService(Context.INPUT_METHOD_SERVICE) as InputMethodManager
      inputMethodManager?.hideSoftInputFromWindow(
        view.windowToken,
        InputMethodManager.HIDE_NOT_ALWAYS
      )

      // 檢察網路連線
      val connMgr = activity!!
        .getSystemService(Context.CONNECTIVITY_SERVICE) as ConnectivityManager
      var networkInfo: NetworkInfo? = null
      if (connMgr != null) {
        networkInfo = connMgr.activeNetworkInfo
      }
      if (networkInfo != null && networkInfo.isConnected) {
        if (isQueryValid(mEditInputTextSearch!!.text)) {
          mTextInputSearch!!.error = null
          mTextInputSearch!!.hint = getString(R.string.query_google_book)
          // Network query
          bookListedViewModel.queryBookListed(mEditInputTextSearch!!.text!!.toString())
        } else {
          mTextInputSearch!!.error = getString(R.string.query_error)
          mTextInputSearch!!.hint = getString(R.string.input_something)
        }
      } else {
        mTextInputSearch!!.error = getString(R.string.error_no_internet)
        mTextInputSearch!!.hint = getString(R.string.error_hint_no_internet)
      }

    }
  }

  private fun isQueryValid(editable: Editable?): Boolean {
    return if (TextUtils.isEmpty(editable)) {
      false
    } else true
  }
}// Required empty public constructor
